contracts/Vote.sol

```diff
@@ -0,0 +1,427 @@
+        return (totalYesVotes, totalNoVotes, totalAbstainVotes);
+    }
+
+    /**
+     * @notice Returns save timestamp of proposal.
+     * @param proposalId The proposal UUID.
+     * @return The timestamp of proposal.
+     */
+    function getProposalTimestamp(uint256 proposalId) public view returns (uint256) {
+        (, , uint256 timestamp, , ) = getGovernance().getProposal(proposalId);
+        return timestamp;
+    }
+
+    /**
+     * @notice Retuns currently locked celo in voting. (This celo cannot be unlocked.)
+     * And it will remove voted proposals from account history if appropriate.
+     * @param beneficiary The beneficiary.
+     */
+    function updateHistoryAndReturnLockedStCeloInVoting(address beneficiary)
+        public
+        onlyManager
+        returns (uint256 lockedAmount)
+    {
+        Voter storage voter = voters[beneficiary];
+
+        uint256 i = voter.votedProposalIds.length;
+        while (i > 0) {
+            uint256 proposalId = voter.votedProposalIds[--i];
+            uint256 proposalTimestamp = proposalTimestamps[proposalId];
+
+            if (proposalTimestamp == 0) {
+                voter.votedProposalIds[i] = voter.votedProposalIds[
+                    voter.votedProposalIds.length - 1
+                ];
+                voter.votedProposalIds.pop();
+                continue;
+            }
+
+            if (block.timestamp < proposalTimestamp + referendumDuration) {
+                VoterRecord storage voterRecord = voter.proposalVotes[proposalId];
+                lockedAmount = Math.max(
+                    lockedAmount,
+                    voterRecord.yesVotes + voterRecord.noVotes + voterRecord.abstainVotes
+                );
+            } else {
+                voter.votedProposalIds[i] = voter.votedProposalIds[
+                    voter.votedProposalIds.length - 1
+                ];
+                voter.votedProposalIds.pop();
+                delete proposalTimestamps[proposalId];
+            }
+        }
+
+        uint256 stCelo = toStakedCelo(lockedAmount);
+        emit LockedStCeloInVoting(beneficiary, stCelo);
+        return stCelo;
+    }
+
+    /**
+     * @notice Retuns proposals still in referendum stage that voter voted on.
+     * @param voter The voter.
+     * @return Proposals in referendum stage.
+     * (For up to date result call updateHistoryAndReturnLockedStCeloInVoting first)
+     */
+    function getVotedStillRelevantProposals(address voter) public view returns (uint256[] memory) {
+        return voters[voter].votedProposalIds;
+    }
+
+    /**
+     * @notice Retuns currently locked celo in voting. (This celo cannot be unlocked.)
+     * @param beneficiary The account.
+     */
+    function getLockedStCeloInVoting(address beneficiary)
+        public
+        view
+        returns (uint256 lockedAmount)
+    {
+        Voter storage voter = voters[beneficiary];
+
+        uint256 i = voter.votedProposalIds.length;
+        while (i > 0) {
```

Comment:
Similar to above:

1. Would be cleaner as a `for` loop.
2. This could run out of gas for a user who hasn't called `updateHistory` in a long time.